test/hotspot/jtreg/serviceability/jvmti/stress/StackTrace/Suspended/libGetStackTraceSuspendedStress.cpp

```diff
@@ -121,6 +121,10 @@ check_vthread_consistency_suspended(jvmtiEnv *jvmti, JNIEnv *jni, jthread vthrea
   //const char* cname = (cthread == NULL) ? "<no cthread>" : get_thread_name(jvmti, jni, cthread);
 
   err = jvmti->SetEventNotificationMode(JVMTI_ENABLE, JVMTI_EVENT_SINGLE_STEP, vthread);
+  if (err == JVMTI_ERROR_THREAD_NOT_ALIVE ||
+      err == JVMTI_ERROR_WRONG_PHASE) {
```

Comment:
WRONG_PHASE looks good to me, but why THREAD_NOT_ALIVE is considered expected for suspended thread?
if the thread was terminated, SuspendThread should return THREAD_NOT_ALIVE, but once SuspendThread returns ERROR_NONE, how the thread can terminates before ResumeThread?

Reply:
My question exactly. I'm not even sure why wrong phase is allowed here.

Reply:
Is the issue here that agent thread started by Debuggee.checkStatus is racing with the test? The producer/consumer thread do 1000 put/take ops and it looks like it can complete and VM commence shutdown while the agent thread is observing, do I read this correctly?

Reply:
David and Alan,
Yes, there is always a race between the app execution and a JVMTI agent. This always was a general issue with the JVMTI. The agent code executing JVMTI functions and events is not protected from the `JVMTI_ERROR_WRONG_PHASE` errors. This error code can be returned by a JVMTI functions even in the middle of the event callback execution. We had this problem in the JDWP agent and fixed it by adding proper synchronization into the agent itself.